<a href="https://colab.research.google.com/github/jcjimenezb123/MNPython-Libro/blob/main/08_RegresionLineal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Regresión Lineal con numpy

In [ ]:
import numpy as np

In [ ]:
x = np.array([0,1,2,3,4,5])
y = np.array([0,0.8,0.9,0.1,-0.8,-1.0])

In [ ]:
pol = np.polyfit(x,y,1)
pol

array([-0.30285714,  0.75714286])

Ecuacion de la linea recta de la regresion lineal

$$
y_{cor}=-0.30285714x +  0.75714286
$$

In [ ]:
y_cor = np.polyval(pol,x)
y_cor

array([ 0.75714286,  0.45428571,  0.15142857, -0.15142857, -0.45428571,
       -0.75714286])

In [ ]:
import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=x,y=y_cor,mode='lines+markers',name='regresion',
                  line=dict(color='royalblue',dash='dot'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion Lineal',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

In [ ]:
sse=sum((y-y_cor)**2)      # variabilidad explicada por la regresion
tse=(len(y))*np.var(y)   # variabilidad total de los datos
r2=1-sse/tse               # medida relativa entre las variabilidades
print(r2)

0.5177880184331798


##Regresion cuadratica

In [ ]:
pol2 = np.polyfit(x,y,deg=2)
pol2

array([-0.16071429,  0.50071429,  0.22142857])

La ecuacion que representa la regresion cuadratica es

$$
y_{cor}=-0.16071429x^2+  0.50071429x+  0.22142857
$$

In [ ]:
y_cor2 = np.polyval(pol2,x)
y_cor2

array([ 0.22142857,  0.56142857,  0.58      ,  0.27714286, -0.34714286,
       -1.29285714])

In [ ]:
sse=sum((y-y_cor2)**2)
tse=(len(y)-1)*np.var(y)
r22=1-sse/tse
print(r22)

0.7946175115207375


In [ ]:
import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=x,y=y_cor2,mode='markers',name='regresion',
                  line=dict(color='royalblue',dash='dash'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion cuadratica',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

In [ ]:
pol3 = np.polyfit(x,y,deg=3)
pol3

array([ 0.08703704, -0.81349206,  1.69312169, -0.03968254])

Usando una regresion cubica tenemos la ecuacion

$$
y_{cor}=0.08703704x^3 -0.81349206x^2+  1.69312169x  -0.03968254
$$

In [ ]:
y_cor3=np.polyval(pol3,x)
y_cor3

array([-0.03968254,  0.92698413,  0.78888889,  0.06825397, -0.71269841,
       -1.03174603])

In [ ]:
sse=sum((y-y_cor3)**2)
tse=(len(y)-1)*np.var(y)
r23=1-sse/tse
print(r23)

0.9846390168970814


In [ ]:
import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=x,y=y_cor3,mode='markers',name='regresion',
                  line=dict(color='royalblue',dash='dash'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion cubica',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

**Problema** En la fabricación de láminas de acero, el método para deformar acero a temperatura normal mantiene una relación inversa con la dureza del
mismo, ya que a medida que la deformación crece, se afecta la dureza del acero. Con el objetivo de conocer ésta relación se tomaron los siguientes datos

$$
\begin{array}{l|c|c|c|c|c|c|c|c|c|}
\text{Deformacion}&6&9&11&13&22&26&28&33&35\\
\hline
\text{Dureza}&68&67&65&53&44&40&37&34&32
\end{array}
$$

La ecuacion que relaciona la dureza con la deformacion es

$$
Dureza=-1.31956215Deformacion+ 75.71998588
$$

0.949561766039208


In [ ]:
import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=x,y=y_cor,mode='lines+markers',name='regresion',
                  line=dict(color='royalblue',dash='dot'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion Lineal',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

**Problema 2** Un proyectil es disparado hacia arriba desde el piso, se mide la altura del proyectil en el tiempo y se obtienen los siguientes datos.

$$
\begin{array}{l|c|c|c|c|c|c|}
\text{Tiempo(s)}& 0& 0.5& 1.0& 1.5& 2.0 &2.5\\
\hline
\text{Altura(ft)}& 0 &20.5& 31.36& 36.25& 30.41& 28.23
\end{array}
$$

array([-12.87142857,  42.22257143,   1.17714286])

La ecuacion del tiro parabolico es

$$
Altura=-12.87142857t^2 + 42.22257143t  + 1.17714286
$$

array([ 1.17714286, 19.07057143, 30.52828571, 35.55028571, 34.13657143,
       26.28714286])

array([ 3.30797943, -0.02764647])

**Problema 3** Con los siguientes datos de Cp del Propano vs Temperatura, obtenga los coeficientes de la ecuación de Cp de la forma.

$$
Cp =a+bT+cT^2+dT^3
$$

Donde $a,b,c,d$ son los coeficientes a determinar y $T$ es la temperatura del Propano.

$$
\begin{array}{l|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|}
\text{Temp(C)}& 50 &100 &150 &200 &273.16 &298.15 &300& 400 &500 &600 &700 &800 &900 &1000 &1100\\
\hline
\text{Cp(kJ/kg)} &34.06 &41.3 &48.79 &56.07 &68.74 &73.6 &73.93 &94.01&112.59& 142.67 &154.77& 163.35 &174.6& 182.67& 128.7
\end{array}
$$

Aplique la regresion cubica para obtener los coeficientes $a,b,c,d.$

array([-1.57390652e-07,  2.00985608e-04,  1.14380718e-01,  2.70935513e+01])

$$
Cp=-1.57390652\times10^{-07}T^3+ 2.00985608\times10^{-04}T^2+1.14380718\times10^{-01}T+  27.0935513
$$

array([ 33.2953774 ,  40.38408853,  48.24164171,  56.74999397,
        70.12663817,  74.89104609,  75.24692374,  94.93053394,
       114.85648064, 134.08041994, 151.65800791, 166.64490065,
       178.09675424, 185.06922476, 186.6179683 ])

En la Regresión Múltiple se tiene más de una variable independiente $x_1,x_2,x_3, ... ,x_n$ y solo una
variable dependiente y. La regresión múltiple nos permite analizar la relación que tiene la variable de
estudio y con las variables regresoras $x_1,x_2,x_3,...,x_n$. Por ejemplo, la velocidad de reacción (variable
de estudio) con la temperatura y la concentración (variables regresoras). Se puede seguir un modelo
lineal

$$
y = a_0+a_1x_1+a_2x_2+...+a_nx_n
$$

Donde los coeficientes a determinar son $a_0,a_1,a_2,...,a_n$.

En un estudio sobre el crecimiento de un par\'asito se contaron los par\'asitos en 10 localizaciones con diversas condiciones de Temperatura y Humedad, dando los siguientes resultados.

$$
\begin{array}{l|c|c|c|c|c|c|c|c|c|c}
T & 15 & 16 & 24 & 13 & 21 & 16 & 22 & 18 & 20 & 16\\
\hline
H & 70 & 65 & 71 & 64 & 84 & 86 & 72 & 84 & 71 & 75\\
\hline
Parasitos & 156 & 157 & 177 & 145 & 197 & 184 & 172 & 187 & 157 & 169
\end{array}
$$

Usar el siguiente modelo para la cantidad de parasitos:

$$
Parasitos=a_1+a_2\times Temperatura+a_3\times Humedad
$$

Estimar los valores de $a_1,a_2$ y $a_3$ de la ecuacion anterior.


In [ ]:
import numpy as np
from scipy.optimize import curve_fit

In [ ]:
x = np.array([[15,16,24,13,21,16,22,18,20,16],\
             [70,65,71,64,84,86,72,84,71,75]])
z = np.array([156,157,177,145,197,184,172,187,157,169])

In [ ]:
def f(x,a1,a2,a3):
    return a1+a2*x[0]+a3*x[1]

In [ ]:
param,_=curve_fit(f,x,z)
print(param)

[13.97853561  1.65787362  1.69964895]


Por lo tanto, la ecuacion es:
$$
Parasitos=13.97853561+1.65787362Temperatura+1.69964895Humedad
$$

In [ ]:
X = np.arange(13,24,0.5)
Y = np.arange(64, 86, 0.5)
X, Y = np.meshgrid(X, Y)
Z=f(np.array([X,Y]),*param)

In [ ]:
import plotly.graph_objects as go

fig = go.Figure(data=[go.Scatter3d(x=x[0],y=x[1],z=z,
                                   mode='markers',
                                   marker=dict(size=5,  color=z),
                                   name='datos'),
                      go.Surface(z=Z,x=X,y=Y,name='regresion')])
fig.update_layout(scene = dict(
                    xaxis_title='Temperatura',
                    yaxis_title='Humedad',
                    zaxis_title='Parasitos'))
fig.show()